python/ray/autoscaler/v2/instance_manager/instance_manager.py

```diff
@@ -180,22 +182,108 @@ def _get_update_im_state_reply(
             reply.status.message = error_message
         return reply
 
-    def _apply_update(self, instance: Instance, update: InstanceUpdateEvent):
+    @staticmethod
+    def _apply_update(instance: Instance, update: InstanceUpdateEvent):
         """
-        Apply the update to the instance.
+        Apply status specific update to the instance.
 
         Args:
             instance: The instance to update.
             update: The update to apply.
-            version: The version of the instance storage.
 
         Raises:
-            InvalidInstanceStatusTransitionError: If the update is invalid.
+            InvalidInstanceUpdateError: If the update is invalid.
         """
-        InstanceUtil.set_status(instance, update.new_instance_status)
+
         if update.new_instance_status == Instance.ALLOCATED:
+            if not update.cloud_instance_id:
+                raise InvalidInstanceUpdateError(
+                    instance_id=update.instance_id,
+                    cur_status=instance.status,
+                    update=update,
+                    details=("ALLOCATED update must have cloud_instance_id"),
+                )
             instance.cloud_instance_id = update.cloud_instance_id
         elif update.new_instance_status == Instance.TERMINATED:
             instance.cloud_instance_id = ""
         elif update.new_instance_status == Instance.RAY_RUNNING:
+            if not update.ray_node_id:
+                raise InvalidInstanceUpdateError(
+                    instance_id=update.instance_id,
+                    cur_status=instance.status,
+                    update=update,
+                    details=("RAY_RUNNING update must have ray_node_id"),
+                )
             instance.node_id = update.ray_node_id
+        elif update.new_instance_status == Instance.REQUESTED:
+            if not update.launch_request_id:
+                raise InvalidInstanceUpdateError(
+                    instance_id=update.instance_id,
+                    cur_status=instance.status,
+                    update=update,
+                    details=("REQUESTED update must have launch_request_id"),
+                )
+            instance.launch_request_id = update.launch_request_id
+
+    @staticmethod
+    def _create_instance(update: InstanceUpdateEvent) -> Instance:
+        """
+        Create a new instance from the given update.
+        """
+
+        if update.new_instance_status not in [
+            # For unmanaged instance not initialized by InstanceManager,
+            # e.g. head node
+            Instance.ALLOCATED,
+            # For new instance being queued to launch.
+            Instance.QUEUED,
+            # For leaked cloud instance that needs to be terminated.
+            Instance.TERMINATING,
+        ]:
+            raise InvalidInstanceUpdateError(
```

Comment:
Should it be assert? How do we handle exception now

Reply:
We will catch exceptions at reconciler, and reconciler would fail the current reconcilation iteration.